runtime/gc_glue_java/FlattenedArrayObjectScanner.hpp

```diff
@@ -0,0 +1,281 @@
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#if !defined(FLATTENEDARRAYOBJECTSCANNER_HPP_)
+#define FLATTENEDARRAYOBJECTSCANNER_HPP_
+
+#include "j9.h"
+#include "j9cfg.h"
+#include "modron.h"
+
+#include "objectdescription.h"
+#include "ArrayObjectModel.hpp"
+#include "GCExtensionsBase.hpp"
+#include "HeadlessMixedObjectScanner.hpp"
+#include "IndexableObjectScanner.hpp"
+
+class GC_FlattenedArrayObjectScanner : public GC_HeadlessMixedObjectScanner
+{
+	/* Data Members */
+private:
+
+	MM_EnvironmentBase *_env;
+	uintptr_t _elementSize; /**< Size of the flattened element, without padding */
+	uintptr_t *_descriptionBasePtr; /**< Pointer to the description base */
+#if defined(OMR_GC_LEAF_BITS)
+	uintptr_t *_leafBasePtr; /**< Pointer to the leaf description base */
+#endif /* defined(OMR_GC_LEAF_BITS) */
+	GC_IndexableObjectScanner _indexableScanner; /**< Used to iterate the array by element */
+
+protected:
+
+public:
+
+	/* Methods */
+private:
+
+protected:
+
+	/**
+	 * @param env The scanning thread environment
+	 * @param[in] arrayPtr pointer to the array to be processed
+	 * @param[in] basePtr pointer to the first contiguous array cell
+	 * @param[in] limitPtr pointer to end of last contiguous array cell
+	 * @param[in] scanPtr pointer to the array cell where scanning will start
+	 * @param[in] endPtr pointer to the array cell where scanning will stop
+	 * @param[in] scanMap The first scan map
+	 * @param[in] elementSize The size of every element
+	 * @param[in] elementClazzPtr Pointer to the element's class
+	 * @param[in] flags Scanning context flags
+	 */
+	MMINLINE GC_FlattenedArrayObjectScanner(
+		MM_EnvironmentBase *env
+		, omrobjectptr_t arrayPtr
+		, fomrobject_t *basePtr
+		, fomrobject_t *limitPtr
+		, fomrobject_t *scanPtr
+		, fomrobject_t *endPtr
+		, uintptr_t elementSize
+		, uintptr_t elementStride
+		, uintptr_t *descriptionBasePtr
+#if defined(OMR_GC_LEAF_BITS)
+		, uintptr_t *leafBasePtr
+#endif /* defined(OMR_GC_LEAF_BITS) */
+		, uintptr_t flags)
+	: GC_HeadlessMixedObjectScanner(env, scanPtr, elementSize, flags | GC_ObjectScanner::indexableObject)
+	, _env(env)
+	, _elementSize(elementSize)
+	, _descriptionBasePtr(descriptionBasePtr)
+#if defined(OMR_GC_LEAF_BITS)
+	, _leafBasePtr(leafBasePtr)
+#endif /* defined(OMR_GC_LEAF_BITS) */
+	, _indexableScanner(env, arrayPtr, basePtr, limitPtr, scanPtr, endPtr, /* scanMap */ 0, elementStride, flags)
```

Comment:
please do not forget to remove commented out

Reply:
Also maybe we need a comment here why `scanMap` is passed 0. It might be confusing. As I understand it never used because we never use Object Scanner from `_indexableScanner` but compare scanPtr and endPtr directly (never check bitmap)

Reply:
This is the correct reason, it is unused.